### DISCUSSION TITLE
Error "type "vecf16" already exists" migrating pgvecto.rs to 0.2.0

### DISCUSSION THREAD

**Discussion by p4block** [immich-app/immich#7310] [Q&A] [answered]
### The bug

This will happen when Immich runs the migration or when using the query described in the new version docs
```
ALTER EXTENSION vectors UPDATE TO '0.2.0';
QueryFailedError: type "vecf16" already exists
```

### The OS that Immich Server is running on
Container

### Version of Immich Server

v1.95.1

### Reproduction steps

```bash
Get into your database, run the migration
```

### Additional information

This is likely to be a pgvecto.rs issue or something specific to my database, leaving here for other people to see.

Database will not be migrated, Immich will not start. 

~~I haven't fully explored more invasive migrations (dropping vectors and recreating the extension for example)~~
Dropping the vectors extension with cascade and restarting may have solved the problem. Remember to take backups of the db.

**Comment by isole** [comment]
I've run into this issue myself, (but with pg16-v0.2.0). I created the missing extension `vectors` before the `ALTER EXTENSION` statement, i.e. 

```
BEGIN;

CREATE SCHEMA IF NOT EXISTS vectors;
ALTER DATABASE immich SET search_path TO "$user", public, vectors;
SET search_path TO "$user", public, vectors;

UPDATE pg_catalog.pg_extension SET extversion = '0.1.11' WHERE extname = 'vectors';
UPDATE pg_catalog.pg_extension SET extrelocatable = true WHERE extname = 'vectors';

CREATE EXTENSION vectors;

ALTER EXTENSION vectors SET SCHEMA vectors;
UPDATE pg_catalog.pg_extension SET extrelocatable = false WHERE extname = 'vectors';
ALTER EXTENSION vectors UPDATE TO '0.2.0';

SELECT pgvectors_upgrade();

COMMIT;
```

Got past the `QueryFailedError`, but I am still getting this:
```
[Nest] 303  - 02/21/2024, 11:43:35 AM     LOG [CommunicationRepository] Initialized websocket server
[Nest] 303  - 02/21/2024, 11:43:35 AM     LOG [DatabaseService] Updating pgvecto.rs extension to 0.2.0
[Nest] 303  - 02/21/2024, 11:43:35 AM    WARN [DatabaseService] 
        The pgvecto.rs extension version is 0.1.11, but 0.2.0 is available.
        Immich attempted to update the extension, but failed to do so.
        This may be because Immich does not have the necessary permissions to update the extension.

        Please run 'ALTER EXTENSION vectors UPDATE' manually as a superuser.
        See https://immich.app/docs/guides/database-queries for how to query the database.
[Nest] 303  - 02/21/2024, 11:43:35 AM   ERROR [DatabaseService] QueryFailedError: permission denied for table pg_extension
/app/immich/server/dist/domain/database/database.service.js:137
            throw new Error(`
                  ^

Error: 
        The pgvecto.rs extension version is 0.1.11, but Immich only supports 0.2.0 and later minor releases.

        If the Postgres instance already has a compatible version installed, Immich may not have the necessary permissions to activate it.
        In this case, please run 'ALTER EXTENSION UPDATE vectors' manually as a superuser.
        See https://immich.app/docs/guides/database-queries for how to query the database.

        Otherwise, please update the version of pgvecto.rs in the Postgres instance to a compatible version.
    at DatabaseService.assertVectorExtension (/app/immich/server/dist/domain/database/database.service.js:137:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /app/immich/server/dist/domain/database/database.service.js:39:13
    at async /app/immich/server/dist/infra/repositories/database.repository.js:185:23
```

Note that the extension vectors is in `0.2.0.`

```
postgres=# alter extension vectors update;
NOTICE:  version "0.2.0" of extension "vectors" is already installed
ALTER EXTENSION
```

**Comment by isole** [comment]
Btw, granting superuser privileges to immich, gets the job done.

`# alter user immich with superuser;`

**Reply by Brian-Graves** [reply]
Not for me. In fact that's the first thing I did when I saw it was the root cause in the Immich logs.

**Comment by andonovski** [comment]
For me, this helped:

```
CREATE SCHEMA IF NOT EXISTS vectors;
UPDATE pg_catalog.pg_extension SET extversion = '0.1.11' WHERE extname = 'vectors';
UPDATE pg_catalog.pg_extension SET extrelocatable = true WHERE extname = 'vectors';

CREATE EXTENSION vectors;

ALTER EXTENSION vectors SET SCHEMA vectors;
UPDATE pg_catalog.pg_extension SET extrelocatable = false WHERE extname = 'vectors';
ALTER EXTENSION vectors UPDATE TO '0.2.0';

SELECT pgvectors_upgrade();
```

It started to work immediately.

**Reply by MrJosu** [reply]
Where you entered this? I am using unraid, this is why I am confused :(

**Comment by JannesV** [comment]
I'm having the same issue I can't seem to upgrade no matter what I try. My immich user is set to superuser but it seems to fail aswell.

**Comment by njzhenghao** [comment]
> For me, this helped:
> 
> ```
> CREATE SCHEMA IF NOT EXISTS vectors;
> UPDATE pg_catalog.pg_extension SET extversion = '0.1.11' WHERE extname = 'vectors';
> UPDATE pg_catalog.pg_extension SET extrelocatable = true WHERE extname = 'vectors';
> 
> CREATE EXTENSION vectors;
> 
> ALTER EXTENSION vectors SET SCHEMA vectors;
> UPDATE pg_catalog.pg_extension SET extrelocatable = false WHERE extname = 'vectors';
> ALTER EXTENSION vectors UPDATE TO '0.2.0';
> 
> SELECT pgvectors_upgrade();
> ```
> 
> It started to work immediately.

could you tell me about your upgrading steps?

for me ,i use this commands to upgrade (docker compose pull && docker compose up -d).
Then the issuse ([Error "type "vecf16" already exists" migrating pgvecto.rs to 0.2.0](https://github.com/immich-app/immich/issues/7302#top))  comes.

**Comment by thehijacker** [comment]
My Immich is also broken. Where do you put those postgres commands?

**Reply by thehijacker** [reply]
Managed to get this far. But still it does not start:

```
psql (14.10 (Debian 14.10-1.pgdg120+1))
Type "help" for help.

postgres=# CREATE SCHEMA IF NOT EXISTS vectors;
NOTICE:  schema "vectors" already exists, skipping
CREATE SCHEMA
postgres=# UPDATE pg_catalog.pg_extension SET extversion = '0.1.11' WHERE extname = 'vectors';
UPDATE 1
postgres=# UPDATE pg_catalog.pg_extension SET extrelocatable = true WHERE extname = 'vectors';
UPDATE 1
postgres=# CREATE EXTENSION vectors;
ERROR:  extension "vectors" already exists
postgres=# ALTER EXTENSION vectors SET SCHEMA vectors;
ALTER EXTENSION
postgres=# UPDATE pg_catalog.pg_extension SET extrelocatable = false WHERE extname = 'vectors';
UPDATE 1
postgres=# ALTER EXTENSION vectors UPDATE TO '0.2.0';
ERROR:  type "vectorindexinfo" does not exist
postgres=# SELECT pgvectors_upgrade();
WARNING:  pgvecto.rs is upgraded. Restart PostgreSQL to take effects.
 pgvectors_upgrade
-------------------

(1 row)

postgres=# ALTER EXTENSION vectors UPDATE;
ERROR:  type "vectorindexinfo" does not exist
```

Log shows:

```
[Nest] 7  - 02/21/2024, 4:09:01 PM    WARN [DatabaseService] 
        The pgvecto.rs extension version is 0.1.11, but 0.2.0 is available.
        Immich attempted to update the extension, but failed to do so.
        This may be because Immich does not have the necessary permissions to update the extension.
        Please run 'ALTER EXTENSION vectors UPDATE' manually as a superuser.
        See https://immich.app/docs/guides/database-queries for how to query the database.
[Nest] 7  - 02/21/2024, 4:09:01 PM   ERROR [DatabaseService] QueryFailedError: type "vecf16" already exists
/usr/src/app/dist/domain/database/database.service.js:137
            throw new Error(`
                  ^
Error: 
        The pgvecto.rs extension version is 0.1.11, but Immich only supports 0.2.0 and later minor releases.
        If the Postgres instance already has a compatible version installed, Immich may not have the necessary permissions to activate it.
        In this case, please run 'ALTER EXTENSION UPDATE vectors' manually as a superuser.
        See https://immich.app/docs/guides/database-queries for how to query the database.
        Otherwise, please update the version of pgvecto.rs in the Postgres instance to a compatible version.
    at DatabaseService.assertVectorExtension (/usr/src/app/dist/domain/database/database.service.js:137:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /usr/src/app/dist/domain/database/database.service.js:39:13
    at async /usr/src/app/dist/infra/repositories/database.repository.js:185:23
Node.js v20.11.0
[Nest] 7  - 02/21/2024, 4:10:03 PM     LOG [CommunicationRepository] Initialized websocket server
[Nest] 7  - 02/21/2024, 4:10:03 PM     LOG [DatabaseService] Updating pgvecto.rs extension to 0.2.0
[Nest] 7  - 02/21/2024, 4:10:03 PM    WARN [DatabaseService] 
        The pgvecto.rs extension version is 0.1.11, but 0.2.0 is available.
        Immich attempted to update the extension, but failed to do so.
        This may be because Immich does not have the necessary permissions to update the extension.
        Please run 'ALTER EXTENSION vectors UPDATE' manually as a superuser.
        See https://immich.app/docs/guides/database-queries for how to query the database.
```

**Reply by thehijacker** [reply]
Making progress. Now it finaly starts. Had to drop extension vectors and recreate it. But when I search I get:

`ERROR [QueryFailedError: column search.embedding does not exist`

**Comment by DavidFederX** [comment]
Hello, 
Same issue here, but no idea where/how to run the postresql commands...
I have a synology with docker and portainer - don't know where to type/paste these commands. 
Could you help?

**Reply by thehijacker** [reply]
See here: https://immich.app/docs/guides/database-queries/

**Reply by ManuelM432** [reply]
when you setup is Unraid with Redis, postgresql14, Immich
And befor in postgresql14 docker -> Repository = tensorchord/pgvecto-rs:pg14-v0.1.11.  change to tensorchord/pgvecto-rs:pg14-v0.2.0
then restart all and it work 
https://github.com/immich-app/immich/releases/tag/v1.95.1

**Reply by DavidFederX** [reply]
Thanks for your replies
Here it's 15, for some reason I don't know. 
I've already replaced old version with v0.2.0 (tensorchord/pgvecto-rs:pg15-v0.2.0)

but I get this: 
`[Nest] 9  - 02/21/2024, 7:28:41 PM    WARN [DatabaseService] 

        The pgvecto.rs extension version is 0.1.14, but 0.2.0 is available.

        Immich attempted to update the extension, but failed to do so.

        This may be because Immich does not have the necessary permissions to update the extension.

        Please run 'ALTER EXTENSION vectors UPDATE' manually as a superuser.

        See https://immich.app/docs/guides/database-queries for how to query the database.

[Nest] 9  - 02/21/2024, 7:28:41 PM   ERROR [DatabaseService] QueryFailedError: extension "vectors" has no update path from version "0.1.14" to version "0.2.0"

/usr/src/app/dist/domain/database/database.service.js:137`

I assume it means immich has no rights on the DB

So I'm trying to do the alter stuff, but have no idea how to do it. I only use GUI for synology, docker, portainer for now - would you know where/how to run commands ?

**Reply by DavidFederX** [reply]
the logs on the DB: 
```
STATEMENT:  ALTER EXTENSION vectors UPDATE TO '0.2.0'
ERROR:  extension "vectors" has no update path from version "0.1.14" to version "0.2.0"
```

**Comment by kenshineto** [comment]
For me running
```
DROP SCHEMA vectors CASCASE;
```

Then running the update commands seemed to work for me.

**Reply by vivek1052** [reply]
How do you do reindexing after dropping and re creating vector schema

**Reply by mertalev** [reply]
Check if you actually have the indexes with this:
```sql
SELECT * FROM pg_stat_all_indexes WHERE indexrelname = 'clip_index';
SELECT * FROM pg_stat_all_indexes WHERE indexrelname = 'face_index';
```
If you don't get a row for a query, then it means you're missing that index and need to re-add it. You can do so with the below query. Please note that this will **remove all smart search and facial recognition data**, so you'll need to re-run those jobs afterwards.
```sql
TRUNCATE TABLE smart_search;
TRUNCATE TABLE asset_faces CASCADE;
TRUNCATE TABLE person;

ALTER TABLE smart_search
ADD COLUMN IF NOT EXISTS embedding vector(512); -- need to change the number here if using a non-default CLIP model

ALTER TABLE asset_faces
ADD COLUMN IF NOT EXISTS embedding vector(512);

SET vectors.pgvector_compatibility=on;

CREATE INDEX IF NOT EXISTS clip_index ON smart_search
    USING hnsw (embedding vector_cosine_ops)
    WITH (ef_construction = 300, m = 16);

CREATE INDEX IF NOT EXISTS face_index ON asset_faces
    USING hnsw (embedding vector_cosine_ops)
    WITH (ef_construction = 300, m = 16);
```

**Reply by qn2060** [reply]
@mertalev 
after operations above, got errors:

server:

[Nest] 7  - 02/23/2024, 9:18:38 AM   ERROR [QueryFailedError: pgvecto.rs: The given vector is invalid for input.
ADVICE: Check if dimensions and scalar type of the vector is matched with the index.
    at PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:219:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async SelectQueryBuilder.loadRawResults (/usr/src/app/node_modules/typeorm/query-builder/SelectQueryBuilder.js:2192:25)
    at async SelectQueryBuilder.executeEntitiesAndRawResults (/usr/src/app/node_modules/typeorm/query-builder/SelectQueryBuilder.js:2040:26)
    at async SelectQueryBuilder.getRawAndEntities (/usr/src/app/node_modules/typeorm/query-builder/SelectQueryBuilder.js:684:29)
    at async SelectQueryBuilder.getMany (/usr/src/app/node_modules/typeorm/query-builder/SelectQueryBuilder.js:750:25)
    at async paginatedBuilder (/usr/src/app/dist/infra/infra.utils.js:49:19)
    at async /usr/src/app/dist/infra/repositories/search.repository.js:74:23
    at async EntityManager.transaction (/usr/src/app/node_modules/typeorm/entity-manager/EntityManager.js:73:28)
    at async SearchRepository.searchSmart (/usr/src/app/dist/infra/repositories/search.repository.js:65:9)] QueryFailedError: pgvecto.rs: The given vector is invalid for input.
ADVICE: Check if dimensions and scalar type of the vector is matched with the index.


postgres:

2024-02-23 07:10:24.733 UTC [1] LOG:  database system is shut down
PostgreSQL Database directory appears to contain a database; Skipping initialization
2024-02-23 07:10:25.167 UTC [1] LOG:  starting PostgreSQL 14.10 (Debian 14.10-1.pgdg120+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 12.2.0-14) 12.2.0, 64-bit
2024-02-23 07:10:25.167 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2024-02-23 07:10:25.167 UTC [1] LOG:  listening on IPv6 address "::", port 5432
2024-02-23 07:10:25.169 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2024-02-23 07:10:25.171 UTC [28] LOG:  database system was shut down at 2024-02-23 07:10:24 UTC
[2024-02-23T07:10:25Z INFO  service::utils::clean] Find directory "pg_vectors/indexes/91898".
[2024-02-23T07:10:25Z INFO  service::utils::clean] Find directory "pg_vectors/indexes/91897".
2024-02-23 07:10:25.176 UTC [1] LOG:  database system is ready to accept connections
2024-02-23 07:11:42.647 UTC [51] ERROR:  pgvecto.rs: The given vector is invalid for input.
	ADVICE: Check if dimensions and scalar type of the vector is matched with the index.
2024-02-23 07:11:42.647 UTC [51] STATEMENT:  SELECT "asset"."id" AS "asset_id", "asset"."deviceAssetId" AS "asset_deviceAssetId", "asset"."ownerId" AS "asset_ownerId", "asset"."libraryId" AS "asset_libraryId", "asset"."deviceId" AS "asset_deviceId", "asset"."type" AS "asset_type", "asset"."originalPath" AS "asset_originalPath", "asset"."resizePath" AS "asset_resizePath", "asset"."webpPath" AS "asset_webpPath", "asset"."thumbhash" AS "asset_thumbhash", "asset"."encodedVideoPath" AS "asset_encodedVideoPath", "asset"."createdAt" AS "asset_createdAt", "asset"."updatedAt" AS "asset_updatedAt", "asset"."deletedAt" AS "asset_deletedAt", "asset"."fileCreatedAt" AS "asset_fileCreatedAt", "asset"."localDateTime" AS "asset_localDateTime", "asset"."fileModifiedAt" AS "asset_fileModifiedAt", "asset"."isFavorite" AS "asset_isFavorite", "asset"."isArchived" AS "asset_isArchived", "asset"."isExternal" AS "asset_isExternal", "asset"."isReadOnly" AS "asset_isReadOnly", "asset"."isOffline" AS "asset_isOffline", "asset"."checksum" AS "asset_checksum", "asset"."duration" AS "asset_duration", "asset"."isVisible" AS "asset_isVisible", "asset"."livePhotoVideoId" AS "asset_livePhotoVideoId", "asset"."originalFileName" AS "asset_originalFileName", "asset"."sidecarPath" AS "asset_sidecarPath", "asset"."stackId" AS "asset_stackId", "exifInfo"."assetId" AS "exifInfo_assetId", "exifInfo"."description" AS "exifInfo_description", "exifInfo"."exifImageWidth" AS "exifInfo_exifImageWidth", "exifInfo"."exifImageHeight" AS "exifInfo_exifImageHeight", "exifInfo"."fileSizeInByte" AS "exifInfo_fileSizeInByte", "exifInfo"."orientation" AS "exifInfo_orientation", "exifInfo"."dateTimeOriginal" AS "exifInfo_dateTimeOriginal", "exifInfo"."modifyDate" AS "exifInfo_modifyDate", "exifInfo"."timeZone" AS "exifInfo_timeZone", "exifInfo"."latitude" AS "exifInfo_latitude", "exifInfo"."longitude" AS "exifInfo_longitude", "exifInfo"."projectionType" AS "exifInfo_projectionType", "exifInfo"."city" AS "exifInfo_city", "exifInfo"."livePhotoCID" AS "exifInfo_livePhotoCID", "exifInfo"."autoStackId" AS "exifInfo_autoStackId", "exifInfo"."state" AS "exifInfo_state", "exifInfo"."country" AS "exifInfo_country", "exifInfo"."make" AS "exifInfo_make", "exifInfo"."model" AS "exifInfo_model", "exifInfo"."lensModel" AS "exifInfo_lensModel", "exifInfo"."fNumber" AS "exifInfo_fNumber", "exifInfo"."focalLength" AS "exifInfo_focalLength", "exifInfo"."iso" AS "exifInfo_iso", "exifInfo"."exposureTime" AS "exifInfo_exposureTime", "exifInfo"."profileDescription" AS "exifInfo_profileDescription", "exifInfo"."colorspace" AS "exifInfo_colorspace", "exifInfo"."bitsPerSample" AS "exifInfo_bitsPerSample", "exifInfo"."fps" AS "exifInfo_fps" FROM "assets" "asset" LEFT JOIN "exif" "exifInfo" ON "exifInfo"."assetId"="asset"."id"  INNER JOIN "smart_search" "search" ON "search"."assetId"="asset"."id" WHERE ( 1=1 AND 1=1 AND 1=1 AND ("asset"."isVisible" = $1 AND "asset"."isArchived" = $2) AND "asset"."ownerId" IN ($3 ) ) AND ( "asset"."deletedAt" IS NULL ) ORDER BY "search"."embedding" <=> $4 ASC LIMIT 101

**Reply by qn2060** [reply]
I used CLIP  XLM-Roberta-Large-Vit-B-16Plus

**Reply by mertalev** [reply]
I believe that model uses a dim size of 640, so try changing the 512 in the `smart_search` command to that

**Reply by qn2060** [reply]
After changing dim size from 512 to 640，my smart search is working now , thanks

**Comment by mertalev** [comment, marked as answer]
This error means the extension has already been upgraded to 0.2.0. The commands in the release notes are only meant to be run before the upgrade. Setting the `extversion` back to 0.1.11 after upgrading makes Immich think it needs to upgrade the extension again, and likewise the extension tries to migrate from 0.1.11. This naturally fails since the extension is really already on 0.2.0.

The fix in this case is just to set `extversion` back to 0.2.0: `UPDATE pg_catalog.pg_extension SET extversion = '0.2.0' WHERE extname = 'vectors';`

**Reply by andrepue** [reply]
after doing all the other commands in the postgresql container before, this finally solve the problem. Thanks a lot

**Comment by chiraanth** [comment]
![Screenshot (7)](https://github.com/immich-app/immich/assets/59216354/2dec3ab2-685c-431d-873b-5a2bdfe6665c)

Any solutions yet?

**Reply by bo0tzz** [reply]
Did you read the release notes before you upgraded?

**Reply by chiraanth** [reply]
I haven't touched it ,it stopped working suddenly 
when i checked the logs i found this

**Reply by bo0tzz** [reply]
Do you use watchtower or another automatic update system?

**Reply by chiraanth** [reply]
Yes i do use watchtower

**Reply by bo0tzz** [reply]
Please don't do that. Immich often has breaking changes that require you to read the release notes before upgrading, which is why you're running into problems now.

**Comment by dajavax** [comment]
If it helps anyone, what solved it for me was to manually run the migration commands posted by @andonovski

> For me, this helped:
> 
> ```
> CREATE SCHEMA IF NOT EXISTS vectors;
> UPDATE pg_catalog.pg_extension SET extversion = '0.1.11' WHERE extname = 'vectors';
> UPDATE pg_catalog.pg_extension SET extrelocatable = true WHERE extname = 'vectors';
> 
> CREATE EXTENSION vectors;
> 
> ALTER EXTENSION vectors SET SCHEMA vectors;
> UPDATE pg_catalog.pg_extension SET extrelocatable = false WHERE extname = 'vectors';
> ALTER EXTENSION vectors UPDATE TO '0.2.0';
> 
> SELECT pgvectors_upgrade();
> ```
> 
> It started to work immediately.

Then restart the containers, and run the extversion update posted by @mertalev 

> This error means the extension has already been upgraded to 0.2.0. The commands in the release notes are only meant to be run before the upgrade. Setting the `extversion` back to 0.1.11 after upgrading makes Immich think it needs to upgrade the extension again, and likewise the extension tries to migrate from 0.1.11. This naturally fails since the extension is really already on 0.2.0.
> 
> The fix in this case is just to set `extversion` back to 0.2.0: `UPDATE pg_catalog.pg_extension SET extversion = '0.2.0' WHERE extname = 'vectors';`

And then restart the containers again.

**Reply by halsafar** [reply]
So for whatever reason the immich postgres container wouldn't let me use `psql` to connect to itself.

I installed psql on the host and was able to connect and the above posted by @dajavax solved it.

```
sudo apt install postgresql-client-common postgresql-client-14
psql -h IMMICH_POSTGRES_CONTAINER_IP --user IMMICH_POSTGRES_USER
# run commands posted
```

**Reply by qn2060** [reply]
> So for whatever reason the immich postgres container wouldn't let me use `psql` to connect to itself.
> 
> I installed psql on the host and was able to connect and the above posted by @dajavax solved it.
> 
> ```
> sudo apt install postgresql-client-common postgresql-client-14
> psql -h IMMICH_POSTGRES_CONTAINER_IP --user IMMICH_POSTGRES_USER
> # run commands posted
> ```


Run the command  `psql -U postgres -d immich `  in the pg container but not host, enter the container use what command  like `docker exec -it mimich-postgres bash`

**Reply by Watever44** [reply]
Thank you ! This really helped me out !

**Comment by Glitch3dPenguin** [comment]
I ran into the same issue. My fault because I didn't read. For anyone else that is coming here because they may have done the same this is the full list of steps to fix. 

**1)** Log into the machine that is hosting Docker (assuming on a Debian based VM) 
```
sudo docker exec -it immich_postgres psql immich USERNAME-IN-.env-HERE
```
**2)** Run the following: 
```
CREATE SCHEMA IF NOT EXISTS vectors;
UPDATE pg_catalog.pg_extension SET extversion = '0.1.11' WHERE extname = 'vectors';
UPDATE pg_catalog.pg_extension SET extrelocatable = true WHERE extname = 'vectors';

CREATE EXTENSION vectors;

ALTER EXTENSION vectors SET SCHEMA vectors;
UPDATE pg_catalog.pg_extension SET extrelocatable = false WHERE extname = 'vectors';
ALTER EXTENSION vectors UPDATE TO '0.2.0';

SELECT pgvectors_upgrade();
\q
```
**3)** Restart all the containers. 
**4)** Run this command again to log into the database container: 
```
sudo docker exec -it immich_postgres psql immich USERNAME-IN-.env-HERE
```
**5)** Run this:
```
UPDATE pg_catalog.pg_extension SET extversion = '0.2.0' WHERE extname = 'vectors'
\q
```
**6)** Restart all the containers again. 

**DO NOT USE WATCHTOWER** to update this stack as it needs your love to update.

**Reply by azuse** [reply]
last step fixed the problem of "type "vecf16" already exists" for me, great
